lib/guardian/topic_guardian.rb

```diff
@@ -199,6 +198,70 @@ def can_see_deleted_topics?(category)
     is_staff? || is_category_group_moderator?(category)
   end
 
+  # Accepts an array of `Topic#id` and returns an array of `Topic#id` which the user can see.
+  def can_see_topic_ids(topic_ids: [], hide_deleted: true)
+    return topic_ids if is_admin?
+    return [] if topic_ids.compact.blank?
+
+    default_scope = Topic.unscoped.where(id: topic_ids)
+
+    # When `hide_deleted` is `true`, hide deleted topics if user is not staff or category moderator
+    if hide_deleted && !is_staff?
+      default_scope = default_scope.where("deleted_at IS NULL")
+
+      if category_group_moderation_enabled?
+        default_scope = default_scope.or(
+          Topic.unscoped.where("deleted_at IS NOT NULL AND topics.category_id IN (#{category_group_moderator_scope.select(:id).to_sql})")
```

Comment:
This results in a query that looks something like:

```sql
WHERE ("topics"."id" IN (…) AND "topics"."deleted_at" IS NULL OR deleted_at IS NOT NULL AND topics.category_id IN (…))
```

which I can't tell if it's correct or not 😄 Doesn't looks right, but maybe it is - `deleted_at` cancel each other out so you end up with id + category_id 🤷

Reply:
This is tricky because there is some lexical precedence at play here. I've wrapped the conditional in brackets so that there is no confusion.